### ISSUE TITLE
Two-level & cumulativity interaction

### ISSUE THREAD

**Issue by ElifUskuplu** [agda/agda#5761]
Hi, I'm making an experiment abou 2LTT, and after adding `two-level` and `cumulativity` flags, I experinced something strange. We can prove Lemma 2.11 in that [paper](https://arxiv.org/abs/1705.03307). However, using a coercion `c : Set i -> SSet i` sending `A |-> A` we can also prove that some maps, which shouldn't be isomorphisms, are isomorphisms. So there might be bug about type checking. Here is a MWE:

```agda
{-# OPTIONS --without-K --two-level --rewriting --cumulativity #-}

open import Agda.Builtin.Equality public
open import Agda.Primitive public
--exo-universe of exotypes
UUᵉ : (i : Level) → SSet (lsuc i)
UUᵉ i = SSet i

--universe of types
UU : (i : Level) → Set (lsuc i)
UU i = Set i

--coercion
c : {i : Level} → UU i → UUᵉ i
c A = A

--exo(strict)equality
data _≡ₛᵉ_ {i : Level} {A : UUᵉ i } (x : A) : A → UUᵉ i where
  refl : x ≡ₛᵉ x

--Id type
Id : {i : Level} {A : UU i} → A → A → UU i
Id {i} {A} = _≡_ {i} {A}

--map from exo-equality to id type
map-≡ : {i : Level} {A : UU i} {u v : A} → u ≡ₛᵉ v → (Id u v)
map-≡ refl = refl

--For the inverse, this works. Here we used "c (Id u v)"
imap-≡1 : {i : Level} {A : UU i} {u v : A} → c (Id u v) → u ≡ₛᵉ v
imap-≡1 refl = refl

--this does not work
imap-≡2 : {i : Level} {A : UU i} {u v : A} → (Id u v) → u ≡ₛᵉ v
imap-≡2 refl = {!!}
--the error:
--Cannot eliminate fibrant type Id u v
--unless target type is also fibrant
--when checking that the pattern refl has type Id u v
```

Here `imap-≡1` shouldn't exist. Similar situations occured for coproduct types, Nat Type, and Empty Type. 
Also, when I use  `two-level` and `cumulativity` together, I'm taking many type check errors. In a long file, this is very unpleasant.

```agda
{-# OPTIONS --without-K --two-level --rewriting  --cumulativity #-}
open import Agda.Builtin.Equality public

--identity type for (fibrant) types
Id : {i : Level} {A : UU i} → A → A → UU i
Id {i} {A} = _≡_ {i} {A}

-- notation for concatenation · is "\ cdot" 
_·_ : {i : Level} {A : UU i} {x y z : A} → Id x y → Id y z → Id x z
refl · q = q

infixr 30 _⁻¹
_⁻¹ : {i : Level} {A : UU i} {x y : A} → Id x y → Id y x
refl ⁻¹ = refl

--Homotopies
_~_ : {i j : Level} {A : UU i} {P : A → UU j} (f g : Π {i} {j} A P) → UU (i ⊔ j)
f ~ g = (x : _) → Id (f x) (g x)

id-htpy : {i j : Level} {A : UU i} {P : A → UU j} (f : Π {i} {j} A P)
          → f ~ f
id-htpy f a = refl

sym-htpy : {i j : Level} {A : UU i} {P : A → UU j} (f g : Π {i} {j} A P)
           → f ~ g → g ~ f
sym-htpy f g H a = (H a) ⁻¹

trns-htpy : {i j : Level} {A : UU i} {P : A → UU j} (f g h : Π {i} {j} A P)
            → f ~ g → g ~ h → f ~ h
trns-htpy f g h H1 H2 a = (H1 a) · (H2 a)
```

Error Message : 

```
———— Errors ————————————————————————————————————————————————
Failed to solve the following constraints:
  j ⊔ _i_42 (f = f) (g = g) (x = a)
    = j ⊔ _i_42 (f = f) (g = h) (x = a)
    (blocked on _i_42)
  j ⊔ _i_42 (f = g) (g = h) (x = a)
    = j ⊔ _i_42 (f = f) (g = g) (x = a)
    (blocked on _i_42)
  j ⊔ _i_42 (f = f) (g = g) (x = a)
    = j ⊔ _i_42 (f = g) (g = f) (x = a)
    (blocked on _i_42)
  piSort (Set i) (λ x → Set (j ⊔ _i_42 (f = f) (g = g) (x = x)))
    =< Set (i ⊔ j)
    (blocked on _i_42)
```

**Comment by ElifUskuplu** [comment]
Hi,

I couldn't handle it myself, but there is a change in the compilation. In the first MWE, `imap-≡1` does not work anymore. I'm taking this error message:
```
Panic: Pattern match failure in do expression at
src/full/Agda/TypeChecking/Primitive/Cubical.hs:2304:7-21
when checking the definition of imap-≡1
```
Since this kind of map shouldn't exist, it's a good news (although I don't understand what happened). However, the pattern matching still seems problematic. Here is a new MWE:
```agda
{-# OPTIONS --without-K --two-level --cumulativity #-}
open import Agda.Builtin.Equality public
open import Agda.Primitive public
--exo-universe of exotypes
UUᵉ : (i : Level) → SSet (lsuc i)
UUᵉ i = SSet i

--universe of types
UU : (i : Level) → Set (lsuc i)
UU i = Set i

--coercion
c : {i : Level} → UU i → UUᵉ i
c A = A

--Naturals
data ℕ : UU lzero where
  zero : ℕ
  succ : ℕ → ℕ

--Exo-naturals
data ℕᵉ : UUᵉ lzero where
  zeroᵉ : ℕᵉ
  succᵉ : ℕᵉ → ℕᵉ

--I expect this is a valid map, and it works.
ℕᵉ-to-cℕ : ℕᵉ → c ℕ
ℕᵉ-to-cℕ zeroᵉ = zero
ℕᵉ-to-cℕ (succᵉ n) = succ (ℕᵉ-to-cℕ n)

--The following does not work, which is good.
ℕ-to-ℕᵉ : ℕ → ℕᵉ
ℕ-to-ℕᵉ n = {!n!}
--I'm taking the following error.
--Cannot eliminate fibrant type ℕ unless target type is also fibrant
--when checking that the expression ? has type ℕᵉ

--However, the following works. This is not something we expected.
cℕ-to-ℕᵉ : c ℕ → ℕᵉ
cℕ-to-ℕᵉ zero = zeroᵉ
cℕ-to-ℕᵉ (succ n) = succᵉ (cℕ-to-ℕᵉ n)
```
I can avoid such a coercion map, but the bug is still there. How can I avoid this?

**Comment by andreasabel** [comment]
> using a coercion `c : Set i -> SSet i` sending `A |-> A`

It seems like there shouldn't be an identity `Set -> SSet` because then I can label any type as non-fibrant and thus circumvent the check _Cannot eliminate fibrant type unless target type is also fibrant_.

So the easiest fix is to remove the subtyping `Set <= SSet`.  @jespercockx: opinion?
https://github.com/agda/agda/blob/6385f850a185523bb5241a71f5a7614a148f6b53/src/full/Agda/TypeChecking/Conversion.hs#L1234-L1240

**Comment by jespercockx** [comment]
As far as I understand, having a subtyping rule `Set <= SSet` is not in general valid for 2-level type theory. Hence it would indeed be better to disable this subtyping.

**Comment by andreasabel** [comment]
Can you get absurdity from eliminating fibrant types to non-fibrant ones?  If yes, such a proof of false would be nice to have here (but I understand too little of the matters to try to construct one myself).

**Comment by jespercockx** [comment]
I guess without any other primitives but `SSet` the answer is no. However, one could for example have a univalent equality in `Set` and a strict equality in `SSet`, in which case casting a type from `Set` to `SSet` would lead to unsoundness.

**Comment by andreasabel** [comment]
> So the easiest fix is to remove the subtyping `Set <= SSet`.

This makes already this definition from the OP ill-typed:
```agda
--map from exo-equality to id type
map-≡ : {i : Level} {A : UU i} {u v : A} → u ≡ₛᵉ v → (Id u v)
map-≡ refl = refl
```
Because `A` is a `Set`, the arguments `u v : A` cannot be subjected to strict equality `u ç v`.  With the subtyping in place, `A` was a possible argument to `≡ₛᵉ`.

This makes me wonder what of 2ltt remains when the subtyping is removed...

Already the initial 2ltt example breaks then: https://github.com/agda/agda/blob/6385f850a185523bb5241a71f5a7614a148f6b53/test/Succeed/SST.agda#L19-L21
```
Set ℓ is not less or equal than SSet ℓ
when checking the definition of _≡ₛ_
```

**Comment by ElifUskuplu** [comment]
As a reference, see https://arxiv.org/pdf/2102.06275.pdf 
In page 12, it's written that "All types are exotypes, but not all exotypes are types." So I think assuming cumulativity, and hence Set <= SSet is not completely harmful. 

In SST.agda file, there are two different strict equalities, one for types, one for exo-types (I mean strict). In my formulation, I assume cumulativity everywhere. I defined strict equality for types in SSet, and so it allows types from Set, too. This is true for all other type formers. I have one for everything and one for only fibrant types. 

Maybe, the problem is the coercion map, because without it, the problem does not occur. Agda does not eliminate fibrant type unless the target type is also fibrant. I thought that the problem shouldn't occur even with the coercion map, but I might be wrong about it.

**Comment by ElifUskuplu** [comment]
I've encountered the problem when I proved Lemma 2.11 in this paper (https://arxiv.org/pdf/1705.03307.pdf ). And here, they don't assume Set <= SSet, but use a coercion C : Set -> SSet. 

I'm not sure but as far as I understand, cumulativity flag gives this coercion secretly. When I also add C map, Agda behaves a fibrant type as a non-fibrant one, so the elimination rule breaks. 

When I don't assume the map C, I can still prove the lemma via cumulativity. However, it still seems to me that even if C, we should avoid such elimination.

**Comment by andreasabel** [comment]
> Maybe, the problem is the coercion map, because without it, the problem does not occur. Agda does not eliminate fibrant type unless the target type is also fibrant. I thought that the problem shouldn't occur even with the coercion map, but I might be wrong about it.

Not sure I understand.  As soon as you have `Set <= SSet`, you have `id : Set -> SSet`.  So if you want cumulativity, you also get the coercion. 
The problem is now that we cannot trust the sort of a type to decide whether a type is fibrant.  Every fibrant type `A : Set` is also non-fibrant, because `A : SSet` as well.  Which sort we store for `A` depends on whether we have applied `id` to `A`, but this is invisible since `id A = A`.  
_The sort of a type_ is no longer a well-defined concept.  We need the _principal sort_ instead.  But Agda hasn't been built with cumulativity in mind, so it does not store the principal sort.

**Comment by ElifUskuplu** [comment]
> Not sure I understand. As soon as you have `Set <= SSet`, you have `id : Set -> SSet`. So if you want cumulativity, you also get the coercion. The problem is now that we cannot trust the sort of a type to decide whether a type is fibrant. Every fibrant type `A : Set` is also non-fibrant, because `A : SSet` as well. Which sort we store for `A` depends on whether we have applied `id` to `A`, but this is invisible since `id A = A`. _The sort of a type_ is no longer a well-defined concept. We need the _principal sort_ instead. But Agda hasn't been built with cumulativity in mind, so it does not store the principal sort.

I mean as in the example : 
```agda
--The following does not work, which is good.
ℕ-to-ℕᵉ : ℕ → ℕᵉ
ℕ-to-ℕᵉ n = {!n!}
--I'm taking the following error.
--Cannot eliminate fibrant type ℕ unless target type is also fibrant
--when checking that the expression ? has type ℕᵉ

--However, the following works. This is not something we expected.
cℕ-to-ℕᵉ : c ℕ → ℕᵉ
cℕ-to-ℕᵉ zero = zeroᵉ
cℕ-to-ℕᵉ (succ n) = succᵉ (cℕ-to-ℕᵉ n)
```
although I assume cumulativity flag, the first kind of maps is not allowed, but the second kind is allowed. That's why I confused about the c map itself.

**Comment by ElifUskuplu** [comment]
> So if you want cumulativity, you also get the coercion.

And that's why I said that, I can avoid to use the coercion map, but the problem is still there.

**Comment by andreasabel** [comment]
I don't have a good idea how to fix this, so I unassign myself.  
This bug happens in the combination of two experimental features, so it has a very low prio for me.  A sledgehammer fix would be to disallow the combination of these features until we have figured out how they should work together...

**Comment by ElifUskuplu** [comment]
I understand, thanks for the discussion.

**Comment by andreasabel** [comment]
I tried to exploit a similar situation for `Prop <= Set`, but there seem to be enough precautions to avoid inconsistency, although it looks pretty dangerous:
```agda
{-# OPTIONS --prop --cumulativity #-}

data   ⊥ : Set where
record ⊤ : Set where
  constructor tt

data Fool : Prop where
  true false : Fool

Bool : Set
Bool = Fool

True : Bool → Set
True true  = ⊤
True false = ⊥

interleaved mutual
  X : Fool → Set

  true≡false : X true → X false
  true≡false x = x

  X = True
```
No it looks like we have `True true -> True false`, but:
```agda
Rohrkrepierer : ⊥
Rohrkrepierer = true≡false tt
```
```
True _ !=< ⊥
when checking that the expression true≡falseT tt has type ⊥
```

Internally, we only have `True (dontcare true) -> True (dontcare false)`...

**Comment by nad** [comment]
Should `--cumulativity` be made incompatible with `--safe`?

**Comment by ElifUskuplu** [comment]
> Should `--cumulativity` be made incompatible with `--safe`?

Agda does not allow both together. When I assume both flags, it says:

`Cannot set OPTIONS pragma --cumulativity with safe flag.`

**Comment by mikeshulman** [comment]
I think the relevant thing to determine whether a given elimination into non-fibrant types is valid should be whether the *inductive definition* that's being eliminated is one that was declared to be fibrant-valued or not.  Maybe this is sort of what you meant by the "principal type", but in my head, at least, it's different.  In the case of a coproduct, for instance, it's not a property of the type `A + B` that we should be looking at, but a property of the type constructor `_+_`.  I don't know anything about the internals of Agda, but it seems to me that that information should be obtainable; don't we have to inspect the inductive definition of `_+_` anyway when doing an elimination, to find out what its constructors are?

**Comment by mikeshulman** [comment]
By the way, I have a concrete reason to need to combine `--two-level` with `--cumulativity`, not just for convenience.  As Elif said, this combination allows defining a single strict equality type that can be applied to both fibrant and non-fibrant types.  But this, in turn, is necessary to allow `--rewriting` to apply to both fibrant and non-fibrant types, since currently only one relation can be declared as `BUILTIN REWRITE`.  (Of course, an alternative fix would be to allow multiple rewriting relations.)

**Comment by mikeshulman** [comment]
(Also by the way, this isn't related to cumulativity, but it appears to be still possible to define a map `⊥ → ⊥ᵉ` from the fibrant empty type to the non-fibrant empty type, without even using any coercions.  This shouldn't be allowed either, by the same principle: you can't eliminate a fibrantly-defined inductive type with a non-fibrant motive.)

**Comment by mikeshulman** [comment]
Regarding the question of consistency: if you can eliminate the fibrant equality type into the non-fibrant one, then you can prove that they're equivalent and hence both satisfy UIP; so this is inconsistent with anything that violates UIP, like an assumption of univalence, or cubical mode which allows to prove univalence.  The analogous eliminations for coproducts, natural numbers, and empty types are consistent with univalence (e.g. they hold in the simplicial model), but still undesirable (in my opinion) since there are other models I care about where they don't hold.

**Comment by ElifUskuplu** [comment]
I was checking the error again. It seems after the fix discussed in https://github.com/agda/agda/issues/7503, the problem is solved. Previously, the following Agda allowed the following
```agda
--However, the following works. This is not something we expected.
cℕ-to-ℕᵉ : c ℕ → ℕᵉ
cℕ-to-ℕᵉ zero = zeroᵉ
cℕ-to-ℕᵉ (succ n) = succᵉ (cℕ-to-ℕᵉ n)
```
But currently (I'm using Agda 2.8.0), it is also not allowed. It gives the expected error:
```agda
Cannot eliminate fibrant type c ℕ
unless target type is also fibrant
when checking that the expression ? has type ℕᵉ
```
Thanks everyone who helped to solve this.
